bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+        } else {
+            let bn = self >> (8 * (size - 3));
+            bn.0.low_u32()
+        };
+
+        if (compact & 0x0080_0000) != 0 {
+            compact >>= 8;
+            size += 1;
+        }
+
+        CompactTarget(compact | (size << 24) as u32)
+    }
+
+    /// Returns true if block hash is less than or equal to this [`Target`].
+    ///
+    /// Proof-of-work validity for a block requires the hash of the block to be less than or equal
+    /// to the target.
+    pub fn is_met_by(&self, hash: BlockHash) -> bool {
+        use crate::hashes::Hash;
+        let hash = U256::from_le_bytes(hash.into_inner());
+        hash <= self.0
+    }
+
+    /// Converts this [`Target`] to [`Work`].
+    ///
+    /// "Work" is defined as the work done to mine a block with this target value (recorded in the
+    /// block header in compact form as nBits). This is not the same as the difficulty to mine a
+    /// block with this target (see `Self::difficulty`).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Target;
+    /// let t = Target::from(0xdeadbeef_u32);
+    /// assert_eq!(t.to_work().to_target(), t)
+    /// ```
+    pub fn to_work(self) -> Work { Work(self.0.inverse()) }
+
+    /// Computes the popular "difficulty" measure for mining.
+    ///
+    /// Difficulty represents how difficult the current target makes it to find a block, relative to
+    /// how difficult it would be at the highest possible target (highest target == lowest difficulty).
+    ///
+    /// For example, a difficulty of 6,695,826 means that at a given hash rate, it will, on average,
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    ///
+    /// # Note
+    ///
+    /// Difficulty is calculated using the following algorithm `max / current` where [max] is
+    /// defined for the Bitcoin network and `current` is the current [target] for this block. As
+    /// such, a low target implies a high difficulty. Since [`Target`] is represented as a 256 bit
+    /// integer but `difficulty()` returns only 128 bits this means for targets below approximately
+    /// `0xffff_ffff_ffff_ffff_ffff_ffff` `difficulty()` will saturate at `u128::MAX`.
+    ///
+    /// [max]: Target::max
+    /// [target]: crate::blockdata::block::BlockHeader::target
+    pub fn difficulty(&self) -> u128 {
+        let d = Target::MAX.0 / self.0;
+        d.saturating_to_u128()
+    }
+}
+do_impl!(Target);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Encoding of 256-bit target as 32-bit float.
+///
+/// This is used to encode a target into the block header. Satoshi made this part of consensus code
+/// in the original version of Bitcoin, likely copying an idea from OpenSSL.
```

Comment:
Not necessarily for now but I'd like to see a more detailed explanation at some point.

Reply:
OpenSSL's bignum (`BN`) type has an encoding, which is even called "compact" as in bitcoin, which is exactly this format.

Reply:
I know what it is, I meant to expand the docs.

Reply:
Add @apoelstra's comment to the code verbatim.

Reply:
Sorry for being unclear, I meant to explain roughly what the format is.